### PR TITLE
Create row struct

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: executor/src/witgen/rows.rs [hunk 1]
```diff
@@ -123,195 +123,188 @@ impl<T: FieldElement> CellValue<T> {
...
-        match val {
-            CellValue::Known(v) => Some(v),
-            _ => None,
-        }
-    }
+#[derive(Clone, Default, PartialEq)]
+pub struct Row<'a, T: FieldElement> {
+    /// The values in the row, zero if unknown.
+    values: WitnessColumnMap<CellValue<T>>,
+    // TODO remove this.
```
Comment (chriseth) [inline executor/src/witgen/rows.rs:173]: Done in #1514

File: executor/src/witgen/rows.rs [hunk 2]
```diff
@@ -123,195 +123,188 @@ impl<T: FieldElement> CellValue<T> {
...
+            .unzip();
+        FinalizedRow::new(values, known_cells)
     }
 }
 
 impl<'a, T: FieldElement> Row<'a, T> {
     /// Creates a "fresh" row, i.e., one that is empty but initialized with the global range constraints.
     pub fn fresh(fixed_data: &'a FixedData<'a, T>, row: RowIndex) -> Row<'a, T> {
-        WitnessColumnMap::from(
+        // TODO this instance could be computed exactly once (per column set) and then cloned.
```
Comment (chriseth) [inline executor/src/witgen/rows.rs:248]: Will do these in a subsequent PR

File: executor/src/witgen/rows.rs [hunk 3]
```diff
@@ -445,26 +445,34 @@ impl<'row, 'a, T: FieldElement> RowPair<'row, 'a, T> {
...
+    }
+
+    fn get_row_mut(&self, next: bool) -> &Row<'a, T> {
+        match next {
+            false => self.current,
+            true => self
+                .next
+                .unwrap_or_else(|| panic!("Tried to access next row, but it is not available.")),
+        }
+    }
```
Comment (georgwiese) [inline executor/src/witgen/rows.rs:459]:
```suggestion
        self.get_row(poly.next).value_is_known(&poly.poly_id)
    }
```
Right?

**Review by georgwiese — COMMENTED** [review]
Cool!

I think the `Default` implementation for `ColumnMap` is broken. In the current code it is only used by `take()` when the row is replaced anyway, but I think the semantics of `Default` should be to give you a valid object. Would there be another solution? I guess we could store `Option<WitnessColumnMap<T>>` in the row, but also kind of annoying.

File: executor/src/witgen/rows.rs [hunk 4]
```diff
@@ -111,10 +110,11 @@ impl std::fmt::Display for RowIndex {
     }
 }
 
-#[derive(Clone, PartialEq, Debug)]
+#[derive(Default, Clone, PartialEq)]
 enum CellValue<T: FieldElement> {
     Known(T),
     RangeConstraint(RangeConstraint<T>),
+    #[default]
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/rows.rs:117]: Me neither, but fortunately, clippy does!

File: executor/src/witgen/rows.rs [hunk 5]
```diff
@@ -123,195 +123,188 @@ impl<T: FieldElement> CellValue<T> {
...
-impl<T: FieldElement> From<CellValue<T>> for Option<T> {
-    fn from(val: CellValue<T>) -> Self {
-        match val {
-            CellValue::Known(v) => Some(v),
-            _ => None,
-        }
-    }
+#[derive(Clone, Default, PartialEq)]
+pub struct Row<'a, T: FieldElement> {
+    /// The values in the row, zero if unknown.
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/rows.rs:171]: True, this was an old comment.

Comment (chriseth) [inline executor/src/witgen/rows.rs:171]:
```suggestion
```

File: executor/src/witgen/rows.rs [hunk 6]
```diff
@@ -123,195 +123,188 @@ impl<T: FieldElement> CellValue<T> {
...
-        self.value.unwrap_or_default()
+    pub fn range_constraint(&self, poly_id: &PolyID) -> Option<RangeConstraint<T>> {
+        match &self.values[poly_id] {
+            CellValue::RangeConstraint(c) => Some(c.clone()),
+            _ => None,
+        }
     }
-}
 
-impl<T: FieldElement> Debug for Cell<'_, T> {
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/rows.rs:190]: Will move it.

**Review by georgwiese — APPROVED** [review]
LGTM

**Pull request merged**